src/transformers/models/propainter/modeling_propainter.py

```diff
@@ -0,0 +1,4583 @@
+                frame_idx = range(0, timesteps)
+                frame_idx = frame_idx[::-1]
+                flow_idx = frame_idx
+                flows_for_prop = flows_forward
+                flows_for_check = flows_backward
+            else:
+                frame_idx = range(0, timesteps)
+                flow_idx = range(-1, timesteps - 1)
+                flows_for_prop = flows_backward
+                flows_for_check = flows_forward
+
+            for i, idx in enumerate(frame_idx):
+                feat_current = features[cache_list[p_i]][idx]
+                mask_current = masks[cache_list[p_i]][idx]
+
+                if i == 0:
+                    feat_prop = feat_current
+                    mask_prop = mask_current
+                else:
+                    flow_prop = flows_for_prop[:, flow_idx[i], :, :, :]
+                    flow_check = flows_for_check[:, flow_idx[i], :, :, :]
+                    flow_vaild_mask = fb_consistency_check(flow_prop, flow_check)
+                    feat_warped = flow_warp(feat_prop, flow_prop.permute(0, 2, 3, 1), interpolation)
+
+                    if self.learnable:
+                        cond = torch.cat(
+                            [
+                                feat_current,
+                                feat_warped,
+                                flow_prop,
+                                flow_vaild_mask,
+                                mask_current,
+                            ],
+                            dim=1,
+                        )
+                        feat_prop = self.deform_align[module_name](feat_prop, cond, flow_prop)
+                        mask_prop = mask_current
+                    else:
+                        mask_prop_valid = flow_warp(mask_prop, flow_prop.permute(0, 2, 3, 1))
+                        mask_prop_valid = torch.where(mask_prop_valid > 0.1, 1, 0).to(mask_prop_valid)
+
+                        union_vaild_mask = mask_current * flow_vaild_mask * (1 - mask_prop_valid)
+                        union_vaild_mask = torch.where(union_vaild_mask > 0.1, 1, 0).to(union_vaild_mask)
+
+                        feat_prop = union_vaild_mask * feat_warped + (1 - union_vaild_mask) * feat_current
+                        # update mask
+                        mask_prop = mask_current * (1 - (flow_vaild_mask * (1 - mask_prop_valid)))
+                        mask_prop = torch.where(mask_prop > 0.1, 1, 0).to(mask_prop)
+
+                # refine
+                if self.learnable:
+                    feat = torch.cat([feat_current, feat_prop, mask_current], dim=1)
+                    feat_prop = feat_prop + self.backbone[module_name](feat)
+
+                features[module_name].append(feat_prop)
+                masks[module_name].append(mask_prop)
+            # end for
+            if "backward" in module_name:
+                features[module_name] = features[module_name][::-1]
+                masks[module_name] = masks[module_name][::-1]
```

Comment:
This core part has to be refactored a bit to match the standards: 
- iteration over backwards/forward needs better naming
- union_vaild... --> union_valid...
- flow_vaild... --> flow_valid...
- what is `cond` ?
- splitting `learnable` or not matches if we are in `train` mode or not? if yes, we should use that instead of a new boolean
- some branching can be reduced here as well

Reply:
- Done in the recent commits
- cond is the conditional features and has been renamed in the recent commits and  is the representation of the context and conditions under which the features are being aligned
- `learnable` parameter is necessary as there are two separate modules being created with it , ie., `img_prop_module` and `feature_propagation_module` [here](https://github.com/huggingface/transformers/blob/ca07202216af8a320218d5e62bf857aace3278a5/src/transformers/models/propainter/modeling_propainter.py#L2147). The image propagation process doesn't involve any learnable operation so `learnable` is used for differentiation among the two modules
- Done in the recent commits